plugins/libplugin-pay.c

```diff
@@ -2637,16 +2673,57 @@ static struct presplit_mod_data *presplit_mod_data_init(struct payment *p)
 
 static u32 payment_max_htlcs(const struct payment *p)
 {
+	const struct payment *root;
 	struct channel_hint *h;
 	u32 res = 0;
 	for (size_t i = 0; i < tal_count(p->channel_hints); i++) {
 		h = &p->channel_hints[i];
 		if (h->local && h->enabled)
 			res += h->htlc_budget;
 	}
+	root = p;
+	while (root->parent)
+		root = root->parent;
+	if (root->max_htlcs && res > *root->max_htlcs)
+		res = *root->max_htlcs;
 	return res;
 }
 
+/** payment_lower_max_htlcs
+ *
+ * @brief indicates that we have a good reason to believe that
+ * we should limit our number of max HTLCs.
+ *
+ * @desc Causes future payment_max_htlcs to have a maximum value
+ * they return.
+ * Can be called by multiple paymods: the lowest one any paymod
+ * has given will be used.
+ * If this is called with a limit higher than the existing limit,
+ * it just successfully returns without doing anything.
+ *
+ * @param p - a payment on the payment tree we should limit.
+ * @param limit - the number of max HTLCs.
+ * @param why - the reason we think the given max HTLCs is
+ * reasonable.
+ */
+static void payment_lower_max_htlcs(struct payment *p, u32 limit,
+				    const char *why)
+{
+	struct payment *root = payment_root(p);
+	if (!root->max_htlcs) {
+		root->max_htlcs = tal(root, u32);
+		*root->max_htlcs = limit;
+		plugin_log(p->plugin, LOG_INFORM,
+			   "Initial limit on max HTLCs: %"PRIu32", %s",
+			   limit, why);
```

Comment:
Just make this a `u32` (not a `u32*`) and initialize it to `UINT32_MAX`, that still maintains the correct semantics without extra allocations.

Reply:
At which point this boils down to `min()` + `plugin_log` and we can just do that in the caller, saving us the indirection of constructing a `why` string and then combine that into a `plugin_log` call.

Reply:
I think converting to `u32` and initializing to `UINT32_MAX` is a good idea, but I disagree with manually inlining the `payment_lower_max_htlcs` function into the sole caller.

The intent is that `payment_lower_max_htlcs` and `payment_max_htlcs` are common, shared parts of the code, that *any* paymod can call.

If we were to manually inline `payment_lower_max_htlcs`, it would imply that the `max_htlcs` field is owned by the `payee_incoming_limit` paymod.  But the intent is that the max-htlcs field is a *shared* field, with a decent monoidal combination (i.e. `min` is a monoid, its zero is `UINT32_MAX`) of multiple paymods poking at it, if in the future we discover even more reasons to limit the number of HTLCs we should emit (and given our recent discoveries, we should probably prepare for other reasons to limit the number of HTLCs).  Thus, rather than hard-inline the min operation, it should be encapsulated away.

Smart compilers will inline anyway.